go/parquet/internal/utils/rle.go

```diff
@@ -0,0 +1,555 @@
+			for i := 0; i < repbatch; i++ {
+				out[i] = r.curVal
+			}
+
+			r.repCount -= int32(repbatch)
+			read += repbatch
+			out = out[repbatch:]
+		} else if r.litCount > 0 {
+			litbatch := int(math.Min(float64(remain), float64(r.litCount)))
+			n, _ := r.r.GetBatch(uint(r.bitWidth), out[:litbatch])
+			if n != litbatch {
+				return read
+			}
+
+			r.litCount -= int32(litbatch)
+			read += litbatch
+			out = out[litbatch:]
+		} else {
+			if !r.Next() {
+				return read
+			}
+		}
+	}
+	return read
+}
+
+func (r *RleDecoder) GetBatchSpaced(vals []uint64, nullcount int, validBits []byte, validBitsOffset int64) (int, error) {
+	if nullcount == 0 {
+		return r.GetBatch(vals), nil
+	}
+
+	converter := plainConverter{}
+	blockCounter := NewBitBlockCounter(validBits, validBitsOffset, int64(len(vals)))
+
+	var (
+		totalProcessed int
+		processed      int
+		block          BitBlockCount
+		err            error
+	)
+
+	for {
+		block = blockCounter.NextFourWords()
+		if block.Len == 0 {
+			break
+		}
+
+		if block.AllSet() {
+			processed = r.GetBatch(vals[:block.Len])
+		} else if block.NoneSet() {
+			converter.FillZero(vals[:block.Len])
+			processed = int(block.Len)
+		} else {
+			processed, err = r.getspaced(converter, vals, int(block.Len), int(block.Len-block.Popcnt), validBits, validBitsOffset)
+			if err != nil {
+				return totalProcessed, err
+			}
+		}
+
+		totalProcessed += processed
+		vals = vals[int(block.Len):]
+		validBitsOffset += int64(block.Len)
+
+		if processed != int(block.Len) {
+			break
+		}
+	}
+	return totalProcessed, nil
+}
+
+func (r *RleDecoder) getspaced(dc DictionaryConverter, vals interface{}, batchSize, nullCount int, validBits []byte, validBitsOffset int64) (int, error) {
+	switch vals := vals.(type) {
+	case []int32:
+		return r.getspacedInt32(dc, vals, batchSize, nullCount, validBits, validBitsOffset)
+	case []int64:
+		return r.getspacedInt64(dc, vals, batchSize, nullCount, validBits, validBitsOffset)
+	case []float32:
+		return r.getspacedFloat32(dc, vals, batchSize, nullCount, validBits, validBitsOffset)
+	case []float64:
+		return r.getspacedFloat64(dc, vals, batchSize, nullCount, validBits, validBitsOffset)
+	case []parquet.ByteArray:
+		return r.getspacedByteArray(dc, vals, batchSize, nullCount, validBits, validBitsOffset)
+	case []parquet.FixedLenByteArray:
+		return r.getspacedFixedLenByteArray(dc, vals, batchSize, nullCount, validBits, validBitsOffset)
+	case []parquet.Int96:
+		return r.getspacedInt96(dc, vals, batchSize, nullCount, validBits, validBitsOffset)
+	case []uint64:
+		if nullCount == batchSize {
+			dc.FillZero(vals[:batchSize])
+			return batchSize, nil
+		}
+
+		read := 0
+		remain := batchSize - nullCount
+
+		const bufferSize = 1024
+		var indexbuffer [bufferSize]IndexType
+
+		// assume no bits to start
+		bitReader := NewBitRunReader(validBits, validBitsOffset, int64(batchSize))
+		validRun := bitReader.NextRun()
+		for read < batchSize {
```

Comment:
is it possible to break this up into different helper methods?

Reply:
I'll take another pass and see if i can further break this up into other helper methods

Reply:
Updated version has the split up into helper functions for this.